pvanalytics/quality/irradiance.py

```diff
@@ -390,3 +392,80 @@ def clearsky_limits(measured, clearsky, csi_max=1.1):
         max_clearsky_index=np.Inf
     )
     return util.check_limits(csi, upper_bound=csi_max, inclusive_upper=True)
+
+
+def _to_hours(freqstr):
+    if freqstr[0].isalpha():
+        freqstr = '1' + freqstr
+    return pd.to_timedelta(freqstr).seconds / 3600
+
+
+def _daily_total(series):
+    # Resample the series returning the integral for each day
+    # calculated using the trapezoid rule.
+    #
+    # If the series has uniform timestamp spacing (i.e. pd.infer_freq
+    # returns a value other than None) then the frequency is used as
+    # the spacing between each value to speed up integration. If the
+    # frequency cannot be inferred, then the hour of the day is
+    # calculated for each value and used as the x-coordinate in the
+    # integration.
+    freq = pd.infer_freq(series.index)
+    if freq:
+        freq_hours = _to_hours(freq)
+        return series.resample('D').apply(
```

Comment:
Only issue I have is not an issue but a request - help me understand how this line of code works :)

Reply:
`resample` give us one groups of all data on each day, we apply `integrate.trapz` to each group. The `dx` argument (or any other kwargs to `apply`) gets passed to `integrate.trapz` as a keyword argument.

Down below (line 424) we have to be more clever because the `x` values change for each group/application. We use a handy trick of calculating all the x values up front, capturing them in a closure, and selecting the right values for the day we are integrating.